prov/verbs_ud/src/verbs_ud_ep.c

```diff
@@ -0,0 +1,807 @@
+{
+	struct ofi_verbs_ud_ep *ep;
+	struct ibv_wc wc = (const struct ibv_wc) { 0 };
+	int ret;
+
+	ep = container_of(util_ep, struct ofi_verbs_ud_ep, util_ep);
+
+	ret = ibv_poll_cq(ep->ibv_shared_cq, 1, &wc);
+	if (ret > 0) {
+		if (likely(wc.status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(
+				(wc.opcode == IBV_WC_SEND) ? util_ep->tx_cq :
+							     util_ep->rx_cq,
+				&wc
+			);
+		else
+			ofi_verbs_ud_cq_report_error(
+				(wc.opcode == IBV_WC_SEND) ? util_ep->tx_cq :
+							     util_ep->rx_cq,
+				&wc
+			);
+	}
+}
+
+static inline
+void ofi_verbs_ud_ep_shared_multiple_progress(struct util_ep *util_ep)
+{
+	struct ofi_verbs_ud_ep *ep;
+	int num_ent, i;
+	struct ibv_wc *wcs;
+
+	ep = container_of(util_ep, struct ofi_verbs_ud_ep, util_ep);
+	wcs = ep->multiple_wcs;
+	
+	num_ent = ibv_poll_cq(ep->ibv_shared_cq,
+			      gl_data.num_poll_multiple_wcs,
+			      wcs);
+	for (i = 0; i < num_ent; i++) {
+		if (likely(wcs[i].status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(
+				(wcs[i].opcode == IBV_WC_SEND) ? util_ep->tx_cq :
+								 util_ep->rx_cq,
+				&wcs[i]
+			);
+		else
+			ofi_verbs_ud_cq_report_error(
+				(wcs[i].opcode == IBV_WC_SEND) ? util_ep->tx_cq :
+								 util_ep->rx_cq,
+				&wcs[i]
+			);
+	}
+}
+
+static inline
+void ofi_verbs_ud_ep_send_recv_progress(struct util_ep *util_ep)
+{
+	struct ofi_verbs_ud_ep *ep;
+	struct ibv_wc wc = (const struct ibv_wc) { 0 };
+	int ret;
+
+	ep = container_of(util_ep, struct ofi_verbs_ud_ep, util_ep);
+
+	/* Poll Transmit events */
+	ret = ibv_poll_cq(ep->ibv_send_cq, 1, &wc);
+	if (ret > 0) {
+		if (likely(wc.status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(util_ep->tx_cq, &wc);
+		else
+			ofi_verbs_ud_cq_report_error(util_ep->tx_cq, &wc);
+	}
+
+	/* Poll Receive events */
+	ret = ibv_poll_cq(ep->ibv_recv_cq, 1, &wc);
+	if (ret > 0) {
+		if (likely(wc.status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(util_ep->rx_cq, &wc);
+		else
+			ofi_verbs_ud_cq_report_error(util_ep->rx_cq, &wc);
+	}
+}
+
+static inline
+void ofi_verbs_ud_ep_send_recv_multiple_progress(struct util_ep *util_ep)
+{
+	struct ofi_verbs_ud_ep *ep;
+	int num_ent, i;
+	struct ibv_wc *wcs;
+
+	ep = container_of(util_ep, struct ofi_verbs_ud_ep, util_ep);
+	wcs = ep->multiple_wcs;
+
+	/* Poll Transmit events */
+	num_ent = ibv_poll_cq(ep->ibv_send_cq,
+			      gl_data.num_poll_multiple_wcs,
+			      wcs);
+	for (i = 0; i < num_ent; i++) {
+		if (likely(wcs[i].status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(util_ep->tx_cq, &wcs[i]);
+		else
+			ofi_verbs_ud_cq_report_error(util_ep->tx_cq, &wcs[i]);
+	}
+
+	/* Poll Receive events */
+	num_ent = ibv_poll_cq(ep->ibv_send_cq,
+			      gl_data.num_poll_multiple_wcs,
+			      wcs);
+	for (i = 0; i < num_ent; i++) {
+		if (likely(wcs[i].status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(util_ep->rx_cq, &wcs[i]);
+		else
+			ofi_verbs_ud_cq_report_error(util_ep->rx_cq, &wcs[i]);
+	}
+}
```

Comment:
The progress functions should be trimmed down to maybe a single call.  I would also consider moving them into the CQ file, since they're more related to cq reading than active endpoint transfers.